Adding two int32 values results in int64, which is wrong [UDTF]

Consider the following example:
```
    @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')
    def redefined_udtf2(x, m, y):
        for i in range(len(x)):
            y[i] = x[i] + np.int32(3)
        return len(x)
```
Running
```
    descr, result = omnisci.sql_execute(f'select * from table(redefined_udtf2(cursor(select i4 from {omnisci.table_name}), 1))')
```
fails with:
```
cannot store i64 to i32*: mismatching types
```

<details>

```
>       descr, result = omnisci.sql_execute(f'select * from table(redefined_udtf2(cursor(select i4 from {omnisci.table_name}), 1))')

rbc/tests/test_omnisci_column_basic.py:385: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
rbc/omniscidb.py:391: in sql_execute
    self.register()
rbc/omniscidb.py:708: in register
    OmnisciCompilerPipeline, self.debug)
rbc/irtools.py:271: in compile_to_LLVM
    pipeline_class=pipeline_class)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:551: in compile_extra
    return pipeline.compile_extra(func)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:331: in compile_extra
    return self._compile_bytecode()
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:393: in _compile_bytecode
    return self._compile_core()
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:373: in _compile_core
    raise e
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:364: in _compile_core
    pm.run(self.state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:347: in run
    raise patched_exception
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:338: in run
    self._runPass(idx, pass_inst, state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_lock.py:32: in _acquire_compile_lock
    return func(*args, **kwargs)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:302: in _runPass
    mutated |= check(pss.run_pass, internal_state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:275: in check
    mangled = func(compiler_state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/typed_passes.py:95: in run_pass
    raise_errors=self._raise_errors)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/typed_passes.py:67: in type_inference_stage
    infer.propagate(raise_errors=raise_errors)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numba.typeinfer.TypeInferer object at 0x7fb13747e890>, raise_errors = True

    def propagate(self, raise_errors=True):
        newtoken = self.get_state_token()
        oldtoken = None
        # Since the number of types are finite, the typesets will eventually
        # stop growing.
    
        while newtoken != oldtoken:
            self.debug.propagate_started()
            oldtoken = newtoken
            # Errors can appear when the type set is incomplete; only
            # raise them when there is no progress anymore.
            errors = self.constraints.propagate(self)
            newtoken = self.get_state_token()
            self.debug.propagate_finished()
        if errors:
            if raise_errors:
                force_lit_args = [e for e in errors
                                  if isinstance(e, ForceLiteralArg)]
                if not force_lit_args:
>                   raise errors[0]
E                   numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E                   Invalid use of Function(<built-in function setitem>) with argument(s) of type(s): (STRUCT__iPLK, int64, int64)
E                    * parameterized
E                   In definition 0:
E                       All templates rejected with literals.
E                   In definition 1:
E                       All templates rejected without literals.
E                   In definition 2:
E                       All templates rejected with literals.
E                   In definition 3:
E                       All templates rejected without literals.
E                   In definition 4:
E                       All templates rejected with literals.
E                   In definition 5:
E                       All templates rejected without literals.
E                   In definition 6:
E                       LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
E                   cannot store i64 to i32*: mismatching types
E                   
E                   File "rbc/omnisci_backend/omnisci_buffer.py", line 219:
E                   def omnisci_buffer_setitem(a, i, v):
E                       <source elided>
E                       if isinstance(a, BufferType):
E                           return lambda a, i, v: omnisci_buffer_setitem_(a, i, v)
E                           ^
E                   
E                   [1] During: lowering "$10call_function.4 = call $2load_global.0(a, i, v, func=$2load_global.0, args=[Var(a, omnisci_buffer.py:219), Var(i, omnisci_buffer.py:219), Var(v, omnisci_buffer.py:219)], kws=(), vararg=None)" at /home/pearu/git/xnd-project/rbc/rbc/omnisci_backend/omnisci_buffer.py (219)
E                       raised from /home/pearu/miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/six.py:669
E                   In definition 7:
E                       LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
E                   cannot store i64 to i32*: mismatching types
E                   
E                   File "rbc/omnisci_backend/omnisci_buffer.py", line 219:
E                   def omnisci_buffer_setitem(a, i, v):
E                       <source elided>
E                       if isinstance(a, BufferType):
E                           return lambda a, i, v: omnisci_buffer_setitem_(a, i, v)
E                           ^
E                   
E                   [1] During: lowering "$10call_function.4 = call $2load_global.0(a, i, v, func=$2load_global.0, args=[Var(a, omnisci_buffer.py:219), Var(i, omnisci_buffer.py:219), Var(v, omnisci_buffer.py:219)], kws=(), vararg=None)" at /home/pearu/git/xnd-project/rbc/rbc/omnisci_backend/omnisci_buffer.py (219)
E                       raised from /home/pearu/miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/six.py:669
E                   In definition 8:
E                       All templates rejected with literals.
E                   In definition 9:
E                       All templates rejected without literals.
E                   In definition 10:
E                       All templates rejected with literals.
E                   In definition 11:
E                       All templates rejected without literals.
E                   This error is usually caused by passing an argument of a type that is unsupported by the named function.
E                   [1] During: typing of setitem at /home/pearu/git/xnd-project/rbc/rbc/tests/test_omnisci_column_basic.py (382)
E                   
E                   File "rbc/tests/test_omnisci_column_basic.py", line 382:
E                       def redefined_udtf2(x, m, y):
E                           <source elided>
E                           for i in range(len(x)):
E                               y[i] = x[i] + np.int32(3)
E                               ^

../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/typeinfer.py:985: TypingError
```

</details>

The following works:
```
    @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')
    def redefined_udtf2(x, m, y):
        for i in range(len(x)):
            y[i] = np.int32(x[i] + 3)
        return len(x)
```

--- a/rbc/omnisci_backend/numpy_funcs.py
+++ b/rbc/omnisci_backend/numpy_funcs.py
@@ -78,8 +78,10 @@ def omnisci_np_zeros(shape, dtype=None):
     else:
         nb_dtype = dtype
 
+    fill_value = False if isinstance(nb_dtype, types.Boolean) else 0
+
     def impl(shape, dtype=None):
-        return full(shape, 0, nb_dtype)  # noqa: F821
+        return full(shape, fill_value, nb_dtype)  # noqa: F821
     return impl
 
 
@@ -91,8 +93,10 @@ def omnisci_np_zeros_like(a, dtype=None):
         else:
             nb_dtype = dtype
 
+        fill_value = False if isinstance(nb_dtype, types.Boolean) else 0
+
         def impl(a, dtype=None):
-            return full_like(a, 0, nb_dtype)  # noqa: F821
+            return full_like(a, fill_value, nb_dtype)  # noqa: F821
         return impl
 
 
@@ -104,8 +108,10 @@ def omnisci_np_ones(shape, dtype=None):
     else:
         nb_dtype = dtype
 
+    fill_value = True if isinstance(nb_dtype, types.Boolean) else 1
+
     def impl(shape, dtype=None):
-        return full(shape, 1, nb_dtype)  # noqa: F821
+        return full(shape, fill_value, nb_dtype)  # noqa: F821
     return impl
 
 
@@ -117,8 +123,10 @@ def omnisci_np_ones_like(a, dtype=None):
         else:
             nb_dtype = dtype
 
+        fill_value = True if isinstance(nb_dtype, types.Boolean) else 1
+
         def impl(a, dtype=None):
-            return full_like(a, 1, nb_dtype)  # noqa: F821
+            return full_like(a, fill_value, nb_dtype)  # noqa: F821
         return impl
 
 

--- a/rbc/omnisci_backend/omnisci_buffer.py
+++ b/rbc/omnisci_backend/omnisci_buffer.py
@@ -185,22 +185,25 @@ def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
     def truncate_or_extend(builder, value, buf_typ):
         # buf[pos] = val
 
-        if isinstance(nb_value, types.Integer):
+        if isinstance(nb_value, types.Integer):  # Integer
             if eltype.bitwidth < nb_value.bitwidth:
-                return builder.trunc(value, buf_typ)
+                return builder.trunc(value, buf_typ)  # truncate
             elif eltype.bitwidth > nb_value.bitwidth:
                 is_signed = nb_value.signed
                 return builder.sext(value, buf_typ) if is_signed else \
-                    builder.zext(value, buf_typ)
-            return value
+                    builder.zext(value, buf_typ)  # extend
         elif isinstance(nb_value, types.Float):  # Floating-point
             if eltype.bitwidth < nb_value.bitwidth:
                 return builder.fptrunc(value, buf_typ)  # truncate
             elif eltype.bitwidth > nb_value.bitwidth:
                 return builder.fpext(value, buf_typ)  # extend
-            return value
-        else:
-            return value
+        elif isinstance(nb_value, types.Boolean):
+            if buf_typ.width < value.type.width:
+                return builder.trunc(value, buf_typ)
+            elif buf_typ.width > value.type.width:
+                return builder.sext(value, buf_typ)
+
+        return value
 
     def codegen(context, builder, signature, args):
         zero = int32_t(0)

--- a/rbc/omnisci_backend/omnisci_buffer.py
+++ b/rbc/omnisci_backend/omnisci_buffer.py
@@ -71,7 +71,7 @@ class BufferPointer(types.Type):
         self.dtype = dtype    # struct dtype
         self.eltype = eltype  # buffer element dtype
         name = "(%s)*" % dtype
-        types.Type.__init__(self, name)
+        super().__init__(name)
 
     @property
     def key(self):
@@ -173,37 +173,38 @@ def omnisci_buffer_getitem(x, i):
         return lambda x, i: omnisci_buffer_getitem_(x, i)
 
 
+# [rbc issue-197] Numba promotes operations like
+# int32(a) + int32(b) to int64
+def truncate_or_extend(builder, nb_value, eltype, value, buf_typ):
+    # buf[pos] = val
+
+    if isinstance(nb_value, types.Integer):  # Integer
+        if eltype.bitwidth < nb_value.bitwidth:
+            return builder.trunc(value, buf_typ)  # truncate
+        elif eltype.bitwidth > nb_value.bitwidth:
+            is_signed = nb_value.signed
+            return builder.sext(value, buf_typ) if is_signed else \
+                builder.zext(value, buf_typ)  # extend
+    elif isinstance(nb_value, types.Float):  # Floating-point
+        if eltype.bitwidth < nb_value.bitwidth:
+            return builder.fptrunc(value, buf_typ)  # truncate
+        elif eltype.bitwidth > nb_value.bitwidth:
+            return builder.fpext(value, buf_typ)  # extend
+    elif isinstance(nb_value, types.Boolean):
+        if buf_typ.width < value.type.width:
+            return builder.trunc(value, buf_typ)
+        elif buf_typ.width > value.type.width:
+            return builder.sext(value, buf_typ)
+
+    return value
+
+
 @extending.intrinsic
 def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
     sig = types.none(data, index, value)
 
     eltype = data.eltype
     nb_value = value
-
-    # [rbc issue-197] Numba promotes operations like
-    # int32(a) + int32(b) to int64
-    def truncate_or_extend(builder, value, buf_typ):
-        # buf[pos] = val
-
-        if isinstance(nb_value, types.Integer):  # Integer
-            if eltype.bitwidth < nb_value.bitwidth:
-                return builder.trunc(value, buf_typ)  # truncate
-            elif eltype.bitwidth > nb_value.bitwidth:
-                is_signed = nb_value.signed
-                return builder.sext(value, buf_typ) if is_signed else \
-                    builder.zext(value, buf_typ)  # extend
-        elif isinstance(nb_value, types.Float):  # Floating-point
-            if eltype.bitwidth < nb_value.bitwidth:
-                return builder.fptrunc(value, buf_typ)  # truncate
-            elif eltype.bitwidth > nb_value.bitwidth:
-                return builder.fpext(value, buf_typ)  # extend
-        elif isinstance(nb_value, types.Boolean):
-            if buf_typ.width < value.type.width:
-                return builder.trunc(value, buf_typ)
-            elif buf_typ.width > value.type.width:
-                return builder.sext(value, buf_typ)
-
-        return value
 
     def codegen(context, builder, signature, args):
         zero = int32_t(0)
@@ -214,7 +215,7 @@ def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
         ptr = builder.load(rawptr)
 
         buf = builder.load(builder.gep(ptr, [zero, zero]))
-        value = truncate_or_extend(builder, value, buf.type.pointee)
+        value = truncate_or_extend(builder, nb_value, eltype, value, buf.type.pointee)
         builder.store(value, builder.gep(buf, [index]))
 
     return sig, codegen
@@ -223,6 +224,9 @@ def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
 @extending.intrinsic
 def omnisci_buffer_setitem_(typingctx, data, index, value):
     sig = types.none(data, index, value)
+
+    eltype = data.members[0].dtype
+    nb_value = value
 
     def codegen(context, builder, signature, args):
         zero = int32_t(0)
@@ -235,6 +239,7 @@ def omnisci_buffer_setitem_(typingctx, data, index, value):
         ptr = builder.load(builder.gep(
             buf, [zero, zero]))
 
+        value = truncate_or_extend(builder, nb_value, eltype, value, ptr.type.pointee)
         builder.store(value, builder.gep(ptr, [index]))
 
     return sig, codegen

--- a/rbc/omnisci_backend/omnisci_buffer.py
+++ b/rbc/omnisci_backend/omnisci_buffer.py
@@ -177,6 +177,29 @@ def omnisci_buffer_getitem(x, i):
 def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
     sig = types.none(data, index, value)
 
+    eltype = data.eltype
+    nb_value = value
+
+    # [rbc issue-197] Numba promotes operations like
+    # int32(a) + int32(b) to int64
+    def truncate_or_extend(builder, value, buf_typ):
+        # buf[pos] = val
+
+        if isinstance(nb_value, types.Integer):
+            if eltype.bitwidth < nb_value.bitwidth:
+                return builder.trunc(value, buf_typ)
+            elif eltype.bitwidth > nb_value.bitwidth:
+                is_signed = nb_value.signed
+                return builder.zext(value, buf_typ) if is_signed else \
+                    builder.sext(value, buf_typ)
+            return value
+        else:  # Floating-point
+            if eltype.bitwidth < nb_value.bitwidth:
+                return builder.fptrunc(value, buf_typ)
+            elif eltype.bitwidth > nb_value.bitwidth:
+                return builder.fpext(value, buf_typ)
+            return value
+
     def codegen(context, builder, signature, args):
         zero = int32_t(0)
 
@@ -186,6 +209,7 @@ def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
         ptr = builder.load(rawptr)
 
         buf = builder.load(builder.gep(ptr, [zero, zero]))
+        value = truncate_or_extend(builder, value, buf.type.pointee)
         builder.store(value, builder.gep(buf, [index]))
 
     return sig, codegen

--- a/rbc/omnisci_backend/omnisci_buffer.py
+++ b/rbc/omnisci_backend/omnisci_buffer.py
@@ -190,14 +190,16 @@ def omnisci_buffer_ptr_setitem_(typingctx, data, index, value):
                 return builder.trunc(value, buf_typ)
             elif eltype.bitwidth > nb_value.bitwidth:
                 is_signed = nb_value.signed
-                return builder.zext(value, buf_typ) if is_signed else \
-                    builder.sext(value, buf_typ)
+                return builder.sext(value, buf_typ) if is_signed else \
+                    builder.zext(value, buf_typ)
             return value
-        else:  # Floating-point
+        elif isinstance(nb_value, types.Float):  # Floating-point
             if eltype.bitwidth < nb_value.bitwidth:
-                return builder.fptrunc(value, buf_typ)
+                return builder.fptrunc(value, buf_typ)  # truncate
             elif eltype.bitwidth > nb_value.bitwidth:
-                return builder.fpext(value, buf_typ)
+                return builder.fpext(value, buf_typ)  # extend
+            return value
+        else:
             return value
 
     def codegen(context, builder, signature, args):

--- a/rbc/omnisci_backend/omnisci_buffer.py
+++ b/rbc/omnisci_backend/omnisci_buffer.py
@@ -58,6 +58,13 @@ class BufferType(types.Type):
     """Numba type class for Omnisci buffer structures.
     """
 
+    @property
+    def eltype(self):
+        """
+        Return buffer element dtype.
+        """
+        return self.members[0].dtype
+
 
 class BufferPointer(types.Type):
     """Numba type class for pointers to Omnisci buffer structures.
@@ -149,7 +156,7 @@ def omnisci_buffer_ptr_getitem_(typingctx, data, index):
 
 @extending.intrinsic
 def omnisci_buffer_getitem_(typingctx, data, index):
-    eltype = data.members[0].dtype
+    eltype = data.eltype
     sig = eltype(data, index)
 
     def codegen(context, builder, signature, args):

--- a/rbc/tests/test_omnisci_array.py
+++ b/rbc/tests/test_omnisci_array.py
@@ -411,9 +411,9 @@ def test_array_constructor_is_null(omnisci):
     assert list(result)[0] == (0,)
 
 
-inps = [('int', 'i4', 'trunc'), ('int', 'i4', 'sext'),
-        ('int', 'i4', 'zext'), ('float', 'f4', 'fptrunc'),
-        ('float', 'f8', 'fpext')]
+inps = [('int32', 'i4', 'trunc'), ('int32', 'i4', 'sext'),
+        ('int32', 'i4', 'zext'), ('float', 'f4', 'fptrunc'),
+        ('double', 'f8', 'fpext')]
 
 
 @pytest.mark.parametrize("typ, col, suffix", inps,
@@ -424,42 +424,24 @@ def test_issue197(omnisci, typ, col, suffix):
     import rbc.omnisci_backend as np
     from numba import types
 
-    @omnisci('int32[](int32[])')
-    def fn_issue197_int_trunc(x):
-        y = np.zeros_like(x)
-        for i in range(len(x)):
-            y[i] = types.int64(x[i] + 3)  # trunc(i64) -> i32
-        return y
+    cast = dict(
+        trunc=types.int64,
+        sext=types.int8,
+        zext=types.uint8,
+        fptrunc=types.float64,
+        fpext=types.float32)[suffix]
 
-    @omnisci('int32[](int32[])')
-    def fn_issue197_int_sext(x):
+    def fn_issue197(x):
         y = np.zeros_like(x)
         for i in range(len(x)):
-            y[i] = types.int8(x[i] + 3)  # sext(i8) -> i32
-        return y
-
-    @omnisci('int32[](int32[])')
-    def fn_issue197_int_zext(x):
-        y = np.zeros_like(x)
-        for i in range(len(x)):
-            y[i] = types.uint8(x[i] + 3)  # zext(i8) -> i32
-        return y
-
-    @omnisci('float[](float[])')
-    def fn_issue197_float_fptrunc(x):
-        y = np.zeros_like(x)
-        for i in range(len(x)):
-            y[i] = types.float64(x[i] + 3.0)  # fptrunc(f64) -> f3232
-        return y
-
-    @omnisci('double[](double[])')
-    def fn_issue197_float_fpext(x):
-        y = np.zeros_like(x)
-        for i in range(len(x)):
-            y[i] = types.float32(x[i] + 3.0)  # fpext(f32) -> f64
+            y[i] = cast(x[i] + 3)
         return y
 
     fn_name = f"fn_issue197_{typ}_{suffix}"
+    fn_issue197.__name__ = fn_name
+
+    omnisci(f'{typ}[]({typ}[])')(fn_issue197)
+
     _, result = omnisci.sql_execute(
         f'SELECT {col}, {fn_name}({col}) FROM {omnisci.table_name};'
     )
@@ -467,3 +449,27 @@ def test_issue197(omnisci, typ, col, suffix):
     column, ret = list(result)[0]
     for x, y in zip(column, ret):
         assert y == x + 3
+
+
+def test_issue197_bool(omnisci):
+    omnisci.reset()
+
+    import rbc.omnisci_backend as np
+
+    @omnisci('bool[](bool[])')
+    def fn_issue197_bool(x):
+        y = np.zeros_like(x)
+        for i in range(len(x)):
+            y[i] = bool(x[i])
+        return y
+
+    col = 'b'
+    fn_name = 'fn_issue197_bool'
+
+    _, result = omnisci.sql_execute(
+        f'SELECT {col}, {fn_name}({col}) FROM {omnisci.table_name};'
+    )
+
+    column, ret = list(result)[0]
+    for x, y in zip(column, ret):
+        assert bool(x) == bool(y)

--- a/rbc/tests/test_omnisci_array.py
+++ b/rbc/tests/test_omnisci_array.py
@@ -409,3 +409,25 @@ def test_array_constructor_is_null(omnisci):
     _, result = omnisci.sql_execute(query)
 
     assert list(result)[0] == (0,)
+
+
+def test_issue197(omnisci):
+    omnisci.reset()
+
+    from rbc.omnisci_backend import Array
+    import numpy as np
+
+    @omnisci('int32[](int32[])')
+    def fn_issue197(x):
+        y = Array(5, 'int32')
+        for i in range(len(x)):
+            y[i] = x[i] + np.int32(3)
+        return y
+
+    _, result = omnisci.sql_execute(
+        f'select i4, fn_issue197(i4) from {omnisci.table_name};'
+    )
+
+    i4, ret = list(result)[0]
+    for x, y in zip(i4, ret):
+        assert y == x + 3

--- a/rbc/tests/test_omnisci_array.py
+++ b/rbc/tests/test_omnisci_array.py
@@ -411,23 +411,59 @@ def test_array_constructor_is_null(omnisci):
     assert list(result)[0] == (0,)
 
 
-def test_issue197(omnisci):
+inps = [('int', 'i4', 'trunc'), ('int', 'i4', 'sext'),
+        ('int', 'i4', 'zext'), ('float', 'f4', 'fptrunc'),
+        ('float', 'f8', 'fpext')]
+
+
+@pytest.mark.parametrize("typ, col, suffix", inps,
+                         ids=[item[-1] for item in inps])
+def test_issue197(omnisci, typ, col, suffix):
     omnisci.reset()
 
-    from rbc.omnisci_backend import Array
-    import numpy as np
+    import rbc.omnisci_backend as np
+    from numba import types
 
     @omnisci('int32[](int32[])')
-    def fn_issue197(x):
-        y = Array(5, 'int32')
+    def fn_issue197_int_trunc(x):
+        y = np.zeros_like(x)
         for i in range(len(x)):
-            y[i] = x[i] + np.int32(3)
+            y[i] = types.int64(x[i] + 3)  # trunc(i64) -> i32
         return y
 
+    @omnisci('int32[](int32[])')
+    def fn_issue197_int_sext(x):
+        y = np.zeros_like(x)
+        for i in range(len(x)):
+            y[i] = types.int8(x[i] + 3)  # sext(i8) -> i32
+        return y
+
+    @omnisci('int32[](int32[])')
+    def fn_issue197_int_zext(x):
+        y = np.zeros_like(x)
+        for i in range(len(x)):
+            y[i] = types.uint8(x[i] + 3)  # zext(i8) -> i32
+        return y
+
+    @omnisci('float[](float[])')
+    def fn_issue197_float_fptrunc(x):
+        y = np.zeros_like(x)
+        for i in range(len(x)):
+            y[i] = types.float64(x[i] + 3.0)  # fptrunc(f64) -> f3232
+        return y
+
+    @omnisci('double[](double[])')
+    def fn_issue197_float_fpext(x):
+        y = np.zeros_like(x)
+        for i in range(len(x)):
+            y[i] = types.float32(x[i] + 3.0)  # fpext(f32) -> f64
+        return y
+
+    fn_name = f"fn_issue197_{typ}_{suffix}"
     _, result = omnisci.sql_execute(
-        f'select i4, fn_issue197(i4) from {omnisci.table_name};'
+        f'SELECT {col}, {fn_name}({col}) FROM {omnisci.table_name};'
     )
 
-    i4, ret = list(result)[0]
-    for x, y in zip(i4, ret):
+    column, ret = list(result)[0]
+    for x, y in zip(column, ret):
         assert y == x + 3

--- a/rbc/tests/test_omnisci_column_basic.py
+++ b/rbc/tests/test_omnisci_column_basic.py
@@ -482,7 +482,7 @@ def test_redefine(omnisci):
     @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')  # noqa: E501, F811
     def redefined_udtf(x, m, y):  # noqa: E501, F811
         for i in range(len(x)):
-            y[i] = x[i] + np.int32(3)
+            y[i] = x[i] + 3
         return len(x)
 
     descr, result = omnisci.sql_execute(

--- a/rbc/tests/test_omnisci_column_basic.py
+++ b/rbc/tests/test_omnisci_column_basic.py
@@ -482,7 +482,7 @@ def test_redefine(omnisci):
     @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')  # noqa: E501, F811
     def redefined_udtf(x, m, y):  # noqa: E501, F811
         for i in range(len(x)):
-            y[i] = np.int32(x[i] + 3)
+            y[i] = x[i] + 3
         return len(x)
 
     descr, result = omnisci.sql_execute(

--- a/rbc/tests/test_omnisci_column_basic.py
+++ b/rbc/tests/test_omnisci_column_basic.py
@@ -482,7 +482,7 @@ def test_redefine(omnisci):
     @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')  # noqa: E501, F811
     def redefined_udtf(x, m, y):  # noqa: E501, F811
         for i in range(len(x)):
-            y[i] = x[i] + 3
+            y[i] = x[i] + np.int32(3)
         return len(x)
 
     descr, result = omnisci.sql_execute(

